### PR TITLE
add extract_fit_time

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This is part of a series of PRs to add a new generic `extract_fit_time()` to allow users to investigate how long different tidymodels objects takes to fit.

TODO:
- [x] make sure names are good.
    - [x] right now `id` clashed with `id` returned from `tune_grid()`
- [x] Make work with non-recipe preprocessor
- [x] add tests
- [x] Make sure this change is backwards compatible
- [ ] remove hardhat from Remotes

Related PRs
https://github.com/tidymodels/hardhat/pull/218
https://github.com/tidymodels/recipes/pull/1071
https://github.com/tidymodels/parsnip/pull/853

``` r
# pak::pak(c(
#   "tidymodels/hardhat@extract_fit_time",
#   "tidymodels/recipes@extract_fit_time",
#   "tidymodels/parsnip@extract_fit_time",
#   "tidymodels/workflows@extract_fit_time"
# ))

library(tidymodels)

data(ames, package = "modeldata")

rec_ames <- recipe(Sale_Price ~ ., data = ames) |>
  step_dummy(all_nominal_predictors()) |>
  step_nzv(all_predictors()) |>
  step_normalize(all_predictors())

lm_spec <- linear_reg()

wf_spec <- workflow(rec_ames, lm_spec)

wf_fit <- fit(wf_spec, ames)

# recipes
wf_fit |>
  extract_recipe() |>
  extract_fit_time()
#> # A tibble: 1 × 2
#>   id      time
#>   <chr>  <dbl>
#> 1 recipe 0.259

wf_fit |>
  extract_recipe() |>
  extract_fit_time(summarize = FALSE)
#> # A tibble: 6 × 2
#>   id                      time
#>   <chr>                  <dbl>
#> 1 prep.dummy_edFdo     0.0190 
#> 2 bake.dummy_edFdo     0.0350 
#> 3 prep.nzv_oSTvE       0.197  
#> 4 bake.nzv_oSTvE       0      
#> 5 prep.normalize_1D9e7 0.00500
#> 6 bake.normalize_1D9e7 0.00300

# parsnip
wf_fit |>
  extract_fit_parsnip() |>
  extract_fit_time() # defaults to summarize = FALSE
#> # A tibble: 1 × 2
#>   id           time
#>   <chr>       <dbl>
#> 1 linear_reg 0.0310

wf_fit |>
  extract_fit_parsnip() |>
  extract_fit_time(summarize = TRUE)
#> Error in `extract_fit_time()`:
#> ! `summarize = TRUE` is not supported for `model_fit` objects.

# workflow
wf_fit |>
  extract_fit_time()
#> # A tibble: 1 × 3
#>   stage    id        time
#>   <chr>    <chr>    <dbl>
#> 1 workflow workflow 0.290

wf_fit |>
  extract_fit_time(summarize = FALSE)
#> # A tibble: 7 × 3
#>   stage      id                      time
#>   <chr>      <chr>                  <dbl>
#> 1 preprocess prep.dummy_edFdo     0.0190 
#> 2 preprocess bake.dummy_edFdo     0.0350 
#> 3 preprocess prep.nzv_oSTvE       0.197  
#> 4 preprocess bake.nzv_oSTvE       0      
#> 5 preprocess prep.normalize_1D9e7 0.00500
#> 6 preprocess bake.normalize_1D9e7 0.00300
#> 7 model      linear_reg           0.0310
```

<sup>Created on 2023-06-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Review by simonpcouch — COMMENTED** [review]
I think the interface is super solid, and the potential uses for this in stacks, workflow sets, and tuning results are many!

Just made a few small docs suggestions—I think, as often as we can, we should use the same term to refer to this value. I think "elapsed fit time" is a good cue, in that it self-documents `system.time()$elapsed` and aligns with the function names. :)

Pending those TODO's, thumbs up from me!

**Comment by simonpcouch** [comment]
re:

> right now `id` clashed with `id` returned from `tune_grid()`

workflowsets uses `wflow_id`, so we have some precedent for distinguishing different types of ids. there’s also `id2` in iterative searches, unfortunately named with a different pattern (though that may be internal-only?). 

I'm not sure I have any silver bullet recommendations. `process_id`? Probably something with `*_id`.

This may be a separate issue from naming, but there may also be some natural linkage to `.config` here. The processing for that will likely happen in tune, but perhaps something worth anticipating.

**Comment by EmilHvitfeldt** [comment]
I like `process_id`!

**Review by hfrick — APPROVED** [review]
I'm wondering if `stage_id` is a good alternative to `process_id`?

File: NEWS.md [hunk 1]
```diff
@@ -10,6 +10,8 @@
 * The prediction columns are now appended to the LHS rather than RHS of 
   `new_data` in `augment.workflow()`, following analogous changes in parsnip (#200).
 
+* New `extract_fit_time()` method has been added that return the time it took to train the workflow. (#191)
```
[the remark this answers is not in the document]
Comment (simonpcouch) [reply NEWS.md:22]: Added in [13e180a](https://github.com/tidymodels/workflows/pull/191/commits/13e180abcec8aef42ede416cf63951a0c215375e).

**Comment by simonpcouch** [comment]
I think a part of this interface that's been tough to wrap my head around is the degree of summarization. I have a hunch that getting summarization right will make deciding on column names a bit easier.

At the moment, output looks like:

``` r
library(tidymodels)

data(ames, package = "modeldata")

rec_ames <- 
  recipe(Sale_Price ~ ., data = ames) |>
  step_dummy(all_nominal_predictors()) |>
  step_nzv(all_predictors()) |>
  step_normalize(all_predictors())

lm_spec <- linear_reg()

wf_fit <- fit(workflow(rec_ames, lm_spec), ames)
wf_fit2 <- fit(workflow(Sale_Price ~ ., lm_spec), ames)

extract_fit_time(wf_fit)
#> # A tibble: 1 × 3
#>   stage    process_id  time
#>   <chr>    <chr>      <dbl>
#> 1 workflow workflow   0.419
extract_fit_time(wf_fit, summarize = FALSE)
#> # A tibble: 7 × 3
#>   stage      process_id              time
#>   <chr>      <chr>                  <dbl>
#> 1 preprocess prep.dummy_wjSqo     0.0290 
#> 2 preprocess bake.dummy_wjSqo     0.156  
#> 3 preprocess prep.nzv_lUZwC       0.199  
#> 4 preprocess bake.nzv_lUZwC       0      
#> 5 preprocess prep.normalize_I9sdp 0.00400
#> 6 preprocess bake.normalize_I9sdp 0.00300
#> 7 model      linear_reg           0.0280

extract_fit_time(wf_fit2)
#> # A tibble: 1 × 3
#>   stage    process_id  time
#>   <chr>    <chr>      <dbl>
#> 1 workflow workflow   0.173
extract_fit_time(wf_fit2, summarize = FALSE)
#> # A tibble: 1 × 3
#>   stage process_id  time
#>   <chr> <chr>      <dbl>
#> 1 model linear_reg 0.173
```

With a recipe, the `summarize` argument reduces the number of rows, but with a formula preprocessor, the only change is the values of `stage` and `process_id`.

Workflows already have the built-in unit ("stage") `pre` (preprocess), `fit` (model), `post` (postprocess). I imagine, in the tuning context, this is likely the unit of observation I'm interesting in timing:

``` r
extract_fit_time(wf_fit, summarize = FALSE) %>%
  group_by(stage) %>%
  summarize(time = sum(time))
#> # A tibble: 2 × 2
#>   stage        time
#>   <chr>       <dbl>
#> 1 model      0.0280
#> 2 preprocess 0.392
```

It'd be nice if that was the default.

An argument like `unit = c("workflow", "stage", "process")` in favor of `summarize` might feel better for workflows. That would make the analogous arguments for the recipe method, though, `"stage"` and `"process"`, which are totally unnatural compared to `summarize = TRUE` or `FALSE` or `unit = "recipe"` or `"step"`. Hmph. 

<sup>Created on 2024-03-27 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by simonpcouch** [comment]
The other remaining question for me is whether we should be thinking of some functionality that measures _predict_ elapsed time alongside this. `extract_predict_time()` is definitely a stranger interface, as the outputs of `bake()` and `predict.model_fit()` are just tibbles. Also, whatever that function is couldn't be passed as an `extract` control argument to resampling functions, because those can only be called on a fitted workflow.

My general leaning is that since 1) the current methods can point out when `bake`ing, at least on `data`, take a notable amount of time and 2) `predict.model_fit()` tends to take much less time than `fit.model_spec()`, we can probably live with just measuring fit time and not providing an analogous interface for prediction.

File: R/extract.R [hunk 2]
```diff
@@ -27,11 +27,15 @@
 #'
 #' - `extract_parameter_set_dials()` returns a set of dials parameter objects.
 #'
+#' - `extract_fit_time()` returns a tibble with elapsed fit times.
```
Comment (simonpcouch) [inline R/extract.R:30]:
```suggestion
#' - `extract_fit_time()` returns a tibble with elapsed fit times. The fit times correspond to
#'   the time for the parsnip engine or recipe steps to fit (or their sum if `summarize = TRUE`)
#'   and do not include other portions of the elapsed time in [workflows::fit.workflow()].
```

Comment (simonpcouch) [reply R/extract.R:30]: Will need a re`document()` and confirmation that that link heads to the right help-page!

**Review by simonpcouch — APPROVED** [review]
Looking solid. :) Just one docs clarification.

**Pull request merged**